src/hash_table.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+from typing import Union
+
+
+class ChainHashTable:
+    class Node:
+        def __init__(self, key: str, val: Union[str, float], next: ChainHashTable.Node = None):
+            self.key = key
+            self.val = val
+            self.next = next
+
+    def __init__(self, capacity: int = 8):
+        self.__initiate(capacity)
+
+    def __initiate(self, capacity):
+        self.__capacity = capacity
+        self.__size = 0
+        self.__bucket = []
+        for _ in range(capacity):
+            self.__bucket.append(self.Node(None, None))
+
+    def capacity(self) -> int:
+        return self.__capacity
+
+    def size(self) -> int:
+        return self.__size
+
+    def insert(self, key: str, val: Union[str, float]):
+        """
+        Insert a new key-value pair to a random hashed bucket.
+        """
+        node: self.Node = self.__bucket[self.__hash(key)]
+        while node.next:
+            node = node.next
+            if node.key == key:
+                node.val = val
+                return
+        node.next = self.Node(key, val)
+        self.__size += 1
+        if self.size() >= self.capacity() / 2:
```

Comment:
I suggest to declaring a constant for `2` or `0.5`. We should not keep magic number in our code. In additional, if we want to change this number in the future, we don't have to look in to the code base if we use constants. Same with line 77.

Reply:
@trungducc got it. I just pushed a new commit to fix this.